File: public/cpp/bot_main.cpp
#include <iostream>
#include <random>
#include <string>
#include <vector>
#include <algorithm>
#include <cstdlib>
#include "lib/base/base_bot.h"
#include "lib/engine/runner.h"
#include "lib/game/game_constants.h"
#include "lib/game/game_state.h"
#include "lib/game/poker_moves.h"
#include "lib/game/round_state.h"
#include "lib/game/terminal_state.h"

class PokerStrategy : public BaseBot {
private:
    std::mt19937 rng;
    std::uniform_real_distribution<double> dist;

public:
    PokerStrategy() : rng(std::random_device()()), dist(0.0, 1.0) {}

    void handleNewRound(const GameState& gameState, const RoundState& roundState, int active) override {
    }

    void handleRoundOver(const GameState& gameState, const TerminalState& terminalState, int active) override {
        const std::array<bool, 2>* bountyHits = terminalState.getBountyHits();
        if (bountyHits && (*bountyHits)[active]) {
            std::string bountyRank = terminalState.getPreviousState()->getBounties()[active];
            std::cout << "Hit bounty of " << bountyRank << "!" << std::endl;
        }
    }

    PokerMove getAction(const GameState& gameState, const RoundState& roundState, int active) override {
        auto legalActions = roundState.getLegalActions();
        int street = roundState.getStreet();
        const auto& myCards = roundState.getHands()[active];
        std::vector<std::string> boardCards(roundState.getDeck().begin(), 
                                           roundState.getDeck().begin() + street);
        int myPip = roundState.getPips()[active];
        int oppPip = roundState.getPips()[1 - active];
        int myStack = roundState.getStacks()[active];
        int oppStack = roundState.getStacks()[1 - active];
        int continueCost = oppPip - myPip;
        std::string myBounty = roundState.getBounties()[active];

        if (legalActions.find(PokerMove::Type::RAISE) != legalActions.end()) {
            auto raiseBounds = roundState.getRaiseBounds();
            if (dist(rng) < 0.4) {
                return RaiseAction(raiseBounds[0]);
            }
        }

        if (legalActions.find(PokerMove::Type::CHECK) != legalActions.end()) {
            return CheckAction();
        }

        if (dist(rng) < 0.2) {
            return FoldAction();
        }

        return CallAction();
    }
};

bool parseArgs(int argc, char* argv[], std::string& host, int& port) {
    host = "localhost";
    port = 0;

    for (int i = 1; i < argc; i++) {
        std::string arg = argv[i];
        if (arg == "--host" && i + 1 < argc) {
            host = argv[++i];
        } else {
            try {
                port = std::stoi(arg);
            } catch (const std::exception& e) {
                std::cerr << "Invalid port: " << arg << std::endl;
                return false;
            }
        }
    }

    if (port == 0) {
        std::cerr << "Port is required" << std::endl;
        return false;
    }

    return true;
}

int main(int argc, char* argv[]) {
    std::string host;
    int port;

    if (!parseArgs(argc, argv, host, port)) {
        return 1;
    }

    PokerStrategy strategy;
    Runner::runBot(&strategy, host, port);
    return 0;
}

File: public/java/src/BotMain.java
import lib.base.BaseBot;
import lib.engine.Runner;
import lib.game.GameConstants;
import lib.game.GameState;
import lib.game.PokerMove;
import lib.game.RoundState;
import lib.game.TerminalState;

import java.util.List;
import java.util.Random;
import java.util.Set;

public class BotMain {
    public static class PokerStrategy implements BaseBot {
        private final Random random = new Random();

        @Override
        public void handleNewRound(GameState gameState, RoundState roundState, int active) {
        }

        @Override
        public void handleRoundOver(GameState gameState, TerminalState terminalState, int active) {
            boolean[] bountyHits = terminalState.getBountyHits();
            if (bountyHits != null && bountyHits[active]) {
                String bountyRank = terminalState.getPreviousState().getBounties()[active];
                System.out.println("Hit bounty of " + bountyRank + "!");
            }
        }

        @Override
        public PokerMove getAction(GameState gameState, RoundState roundState, int active) {
            Set<PokerMove> legalActions = roundState.getLegalActions();
            int street = roundState.getStreet();
            List<String> myCards = roundState.getHands()[active];
            List<String> boardCards = roundState.getDeck().subList(0, street);
            int myPip = roundState.getPips()[active];
            int oppPip = roundState.getPips()[1 - active];
            int myStack = roundState.getStacks()[active];
            int oppStack = roundState.getStacks()[1 - active];
            int continueCost = oppPip - myPip;
            String myBounty = roundState.getBounties()[active];

            for (PokerMove action : legalActions) {
                if (action instanceof PokerMove.Raise) {
                    int[] raiseBounds = roundState.getRaiseBounds();
                    if (random.nextDouble() < 0.4) {
                        return PokerMove.raise(raiseBounds[0]);
                    }
                }
            }

            if (legalActions.contains(PokerMove.check())) {
                return PokerMove.check();
            }

            if (random.nextDouble() < 0.2) {
                return PokerMove.fold();
            }

            return PokerMove.call();
        }
    }

    public static void main(String[] args) {
        String host = "localhost";
        int port = 0;

        for (int i = 0; i < args.length; i++) {
            if (args[i].equals("--host") && i + 1 < args.length) {
                host = args[i + 1];
                i++;
            } else {
                try {
                    port = Integer.parseInt(args[i]);
                } catch (NumberFormatException e) {
                    System.err.println("Invalid port: " + args[i]);
                    System.exit(1);
                }
            }
        }

        if (port == 0) {
            System.err.println("Port is required");
            System.exit(1);
        }

        Runner.runBot(new PokerStrategy(), host, port);
    }
}